java/src/main/kotlin/io/spine/validation/java/setonce/SetOnceConstraintViolation.kt

```diff
@@ -0,0 +1,167 @@
+import io.spine.validation.java.setonce.MessageToken.FIELD_NAME
+import io.spine.validation.java.setonce.MessageToken.PROPOSED_VALUE
+import io.spine.validation.java.setonce.MessageToken.values
+
+/**
+ * Builds a [ConstraintViolation] instance for the given field.
+ *
+ * @param errorMessage The error message pattern.
+ * @param field The field, which was attempted to set twice.
+ */
+internal class SetOnceConstraintViolation(
+    private val errorMessage: String,
+    field: Field
+) {
+
+    private val fieldName = field.name.value
+    private val qualifiedName = field.qualifiedName
+    private val declaringMessage = field.declaringType
+
+    /**
+     * Builds an expression that returns a new instance of [ConstraintViolation]
+     * with the given values.
+     *
+     * @param currentValue The current field value as string.
+     * @param newValue The proposed new value as string.
+     * @param payload The violated value to be packed as Protobuf `Any` and set for
+     *  [ConstraintViolation.getFieldValue] property. Usually, it is just [newValue].
+     *  But for some field types, some conversion may take place.
+     */
+    fun withValues(
+        currentValue: Expression<String>,
+        newValue: Expression<String>,
+        payload: Expression<*> = newValue
+    ): Expression<ConstraintViolation> {
+        val tokenValues = tokenValues(currentValue, newValue)
+        val (format, params) = toPrintfString(errorMessage, tokenValues)
+        val fieldPath = ClassName(FieldPath::class).newBuilder()
+            .chainAdd("field_name", StringLiteral(fieldName))
+            .chainBuild<FieldPath>()
+        val violation = ClassName(ConstraintViolation::class).newBuilder()
+            .chainSet("msg_format", StringLiteral(format))
+            .chainAddAll("param", listExpression(params))
+            .chainSet("type_name", StringLiteral(declaringMessage.qualifiedName))
+            .chainSet("field_path", fieldPath)
+            .chainSet("field_value", payload.packToAny())
+            .chainBuild<ConstraintViolation>()
+        return violation
+    }
+
+    /**
+     * Determines the value for each of the supported tokens.
+     */
+    private fun tokenValues(currentValue: Expression<String>, newValue: Expression<String>) =
+        MessageToken.values().associateWith {
+            when (it) {
+                FIELD_NAME -> StringLiteral(fieldName)
+                CURRENT_VALUE -> currentValue
+                PROPOSED_VALUE -> newValue
+            }
+        }
+
+    /**
+     * Prepares `printf`-style format string and its parameters.
+     *
+     * This method does the following:
+     *
+     * 1. Finds all [MessageToken]s in the given [errorMessage] and replaces them with `%s`.
+     * 2. For each found token, it finds its corresponding [value][tokenValues], and puts it
+     *    to the list of parameters.
+     */
+    private fun toPrintfString(
```

Comment:
I would transform it to at least internal thing, at cover with tests separately. It's an important piece.

Reply:
Moving it up now will conflict with already present classes that help to assemble `ConstraintViolation`, though they do it in a bit different way. I've created an issue for this: https://github.com/SpineEventEngine/validation/issues/167